agimus_controller/agimus_controller/ocp_base_croco.py

```diff
@@ -0,0 +1,226 @@
+                )
+                # Create Differential Action Model (DAM), i.e. continuous dynamics and cost functions
+                self._running_DAM = crocoddyl.DifferentialActionModelFreeFwdDynamics(
+                    self._state,
+                    self._actuation,
+                    runningCostModel,
+                )
+                # Create Integrated Action Model (IAM), i.e. Euler integration of continuous dynamics and cost
+                runningModel = crocoddyl.IntegratedActionModelEuler(
+                    self._running_DAM, self.dt
+                )
+                runningModel.differential.armature = OCPParamsCrocoBase.armature
+                self._runningModelList.append(runningModel)
+            else:
+                terminalCostModel = crocoddyl.CostModelSum(self._state)
+                terminalCostModel.addCost(
+                    "stateReg",
+                    xRegCost,
+                    np.concatenate(
+                        [
+                            self._OCPParams.WeightedTrajectoryPoints[
+                                t
+                            ].weight.w_robot_configuration,
+                            self._OCPParams.WeightedTrajectoryPoints[
+                                t
+                            ].weight.w_robot_velocity,
+                        ]
+                    ),
+                )
+                terminalCostModel.addCost(
+                    "gripperPose",
+                    goalTrackingCost,
+                    self._OCPParams.WeightedTrajectoryPoints[
+                        t
+                    ].weight.w_end_effector_poses,
+                )
+
+                self._terminal_DAM = crocoddyl.DifferentialActionModelFreeFwdDynamics(
+                    self._state, self._actuation, terminalCostModel
+                )
+
+                self._terminalModel = crocoddyl.IntegratedActionModelEuler(
+                    self._terminal_DAM, 0.0
+                )
+                self._terminalModel.differential.armature = OCPParamsCrocoBase.armature
+
+        problem = crocoddyl.ShootingProblem(
+            self.x0, self._runningModelList, self._terminalModel
+        )
+        # Create solver + callbacks
+        ocp = mim_solvers.SolverSQP(problem)
+
+        # Merit function
+        ocp.use_filter_line_search = False
+
+        # Parameters of the solver
+        ocp.termination_tolerance = OCPParamsCrocoBase.termination_tolerance
+        ocp.max_qp_iters = OCPParamsCrocoBase.qp_iters
+        ocp.eps_abs = OCPParamsCrocoBase.eps_abs
+        ocp.eps_rel = OCPParamsCrocoBase.eps_rel
+
+        ocp.with_callbacks = OCPParamsCrocoBase.callbacks
+
+        result = ocp.solve(self.x_init, self.u_init, OCPParamsCrocoBase.solver_iters)
+
+        self.ocp_results = OCPResults(
+            states=ocp.xs,
+            ricatti_gains=ocp.Ks,  # Not sure about this one
+            feed_forward_terms=ocp.us,
```

Comment:
should we return the whole trajectory in the debug result class, and simply like x0, u0 and k0 in the result?

Reply:
for the `OCPResults` I think that's what it should be because that's what  we need for the LFC message, but I'm seeing that the type hinting of `OCPResults` suggests we would give the whole arrays of xs, us and K, maybe that's a mistake ?

Reply:
Yes we should return the full trajectory.
There are ongoing discussion on the low level interface for the communication between the MPC and the LFC.
In python we don't care much as no data are actually copied here.
So we better return as much knowledge as possible ans use whatever in ROS behind.